File: aware-accelerometer/src/androidTest/java/com/aware/android/sensor/accelerometer/DatabaseTest.java
package com.aware.android.sensor.accelerometer;

import android.content.Context;
import android.hardware.Sensor;
import android.hardware.SensorManager;
import android.support.test.InstrumentationRegistry;
import android.support.test.runner.AndroidJUnit4;

import com.aware.android.sensor.accelerometer.db.DbEngine;
import com.aware.android.sensor.accelerometer.model.AccelerometerDevice;
import com.aware.android.sensor.accelerometer.model.AccelerometerEvent;
import com.aware.android.sensor.core.db.Engine;

import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import static org.junit.Assert.*;

/**
 * Instrumented test, which will execute on an Android device.
 *
 * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
 */
@RunWith(AndroidJUnit4.class)
public class DatabaseTest {

    static final String ENCRYPTION_KEY = "custom_key";

    @Test
    public void initDatabases() throws Exception {
        // Context of the app under test.
        Context appContext = InstrumentationRegistry.getTargetContext();

        Engine engine = new DbEngine.Builder(appContext)
                .setDatabaseType(Engine.DatabaseType.NONE)
                .build();

        assertNull(engine);

        engine = new DbEngine.Builder(appContext)
                .setDatabaseName("test.db")
                .setDatabaseType(Engine.DatabaseType.ROOM)
                .build();

        assertNotNull(engine);

        engine = new DbEngine.Builder(appContext)
                .setDatabaseName("test_encrypted.db")
                .setDatabaseType(Engine.DatabaseType.ROOM)
                .setEncryptionKey(ENCRYPTION_KEY)
                .build();

        assertNotNull(engine);
    }

    @Test
    public void testInsertAllEvents() throws Exception {
        // Context of the app under test.
        Context appContext = InstrumentationRegistry.getTargetContext();
        // Create some default events
        ArrayList<AccelerometerEvent> events = new ArrayList<>();
        for (int i = 0; i < 100; i++) {
            events.add(new AccelerometerEvent());
        }

        final AccelerometerEvent[] data_buffer = new AccelerometerEvent[events.size()];
        events.toArray(data_buffer);

        Engine engine = new DbEngine.Builder(appContext)
                .setDatabaseName("test.db")
                .setDatabaseType(Engine.DatabaseType.ROOM)
                .build();
        engine.removeAll().join();
        engine.save(data_buffer).join();
        List<AccelerometerEvent> data = engine.getAll(AccelerometerEvent.class);
        assertEquals(events.size(), data.size());
        engine.close();

        Engine encryptedEngine = new DbEngine.Builder(appContext)
                .setDatabaseName("test_encrypted.db")
                .setDatabaseType(Engine.DatabaseType.ROOM)
                .setEncryptionKey(ENCRYPTION_KEY)
                .build();
        encryptedEngine.removeAll().join();
        encryptedEngine.save(data_buffer).join();
        List<AccelerometerEvent> data2 = encryptedEngine.getAll(AccelerometerEvent.class);
        assertEquals(events.size(), data2.size());
        encryptedEngine.close();
    }

    @Test
    public void testUpdateDevice() throws Exception {
        // Context of the app under test.
        Context appContext = InstrumentationRegistry.getTargetContext();


        SensorManager mSensorManager = (SensorManager) appContext.getSystemService(Context.SENSOR_SERVICE);
        Sensor mAccelerometer = mSensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER);

        AccelerometerDevice device = new AccelerometerDevice("", System.currentTimeMillis(), mAccelerometer);

        Engine engine = new DbEngine.Builder(appContext)
                .setDatabaseName("test.db")
                .setDatabaseType(Engine.DatabaseType.ROOM)
                .build();
        engine.removeAll().join();
        engine.save(device).join();

        List<AccelerometerDevice> savedDevice = engine.getAll(AccelerometerDevice.class);
        assertEquals(1, savedDevice.size());
        // TODO (sercant): we should also check if the entries are same here.

        String uuid = UUID.randomUUID().toString();
        device.setDeviceId(uuid);
        engine.save(device).join();

        savedDevice = engine.getAll(AccelerometerDevice.class);
        assertEquals(1, savedDevice.size());
        assertEquals(device.getDeviceId(), savedDevice.get(0).getDeviceId());
    }
}

File: library/src/androidTest/java/com/awareframework/android/core/DatabaseTest.kt
package com.awareframework.android.core

import android.support.test.InstrumentationRegistry
import android.support.test.runner.AndroidJUnit4
import com.awareframework.android.core.db.Engine
import com.awareframework.android.core.model.AwareData
import com.awareframework.android.core.model.AwareObject
import junit.framework.TestCase.*
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import java.util.*

/**
 * Class decription
 *
 * @author  sercant
 * @date 23/03/2018
 */
@RunWith(AndroidJUnit4::class)
class DatabaseTest {

    private var engine: Engine? = null
    private var encryptedEngine: Engine? = null

    val TABLE_NAME_MULTIPLE_ENTRY = "testTableMultiple"
    val TABLE_NAME_SINGLE_ENTRY = "testTableSingle"

    @Before
    @Throws(Exception::class)
    fun init() {
        // Context of the app under test.
        val appContext = InstrumentationRegistry.getTargetContext()

        engine = Engine.Builder(appContext)
                .setType(Engine.DatabaseType.NONE)
                .build()

        assertNull(engine)

        engine = Engine.Builder(appContext)
                .setPath("test.db")
                .setType(Engine.DatabaseType.ROOM)
                .build()

        assertNotNull(engine)

        encryptedEngine = Engine.Builder(appContext)
                .setPath("test_encrypted.db")
                .setType(Engine.DatabaseType.ROOM)
                .setEncryptionKey(ENCRYPTION_KEY)
                .build()

        assertNotNull(engine)
    }

    @Test
    @Throws(Exception::class)
    fun testDeleteData() {
        val engine = engine!!

        // Create some default events
        val events = ArrayList<AwareObject>()
        for (i in 0..99) {
            events.add(AwareObject())
        }

        val data_buffer: Array<AwareObject> = events.toTypedArray()
        engine.removeAll().join()
        engine.save(data_buffer, TABLE_NAME_MULTIPLE_ENTRY).join()

        var data: List<AwareData>? = null
        engine.get(TABLE_NAME_MULTIPLE_ENTRY, 10, { data = it }).join()

        assertEquals(10, data!!.size.toLong())

        engine.remove(data!!).join()

        var afterRemoveData: List<AwareData>? = null
        engine.get(TABLE_NAME_MULTIPLE_ENTRY, 100, { afterRemoveData = it }).join()

        assertEquals(90, afterRemoveData!!.size.toLong())
    }

    @Test
    @Throws(Exception::class)
    fun testInsertAllEvents() {
        // Context of the app under test.
        val appContext = InstrumentationRegistry.getTargetContext()
        // Create some default events
        val events = ArrayList<AwareObject>()
        for (i in 0..99) {
            events.add(AwareObject())
        }

        val data_buffer: Array<AwareObject> = events.toTypedArray()

        engine!!.removeAll().join()
        engine!!.save(data_buffer, TABLE_NAME_MULTIPLE_ENTRY).join()
        var data: List<AwareData>? = null
        engine!!.getAll(TABLE_NAME_MULTIPLE_ENTRY, { data = it }).join()
        assertEquals(events.size.toLong(), data!!.size.toLong())

        encryptedEngine!!.removeAll().join()
        encryptedEngine!!.save(data_buffer, TABLE_NAME_MULTIPLE_ENTRY).join()
        var data2: List<AwareData>? = null
        encryptedEngine!!.getAll(TABLE_NAME_MULTIPLE_ENTRY, { data2 = it }).join()
        assertEquals(events.size.toLong(), data2!!.size.toLong())
    }

    @Test
    @Throws(Exception::class)
    fun testUpdateDevice() {
        val device = AwareObject()

        engine!!.removeAll().join()
        engine!!.save(device, TABLE_NAME_SINGLE_ENTRY, 0L).join()

        var savedDevice: List<AwareData>? = null
        engine!!.getAll(TABLE_NAME_SINGLE_ENTRY, { savedDevice = it }).join()

        assertEquals(1, savedDevice!!.size.toLong())

        val device2 = savedDevice!![0]

        device2.alterData<AwareObject> {
            val uuid = UUID.randomUUID().toString()
            it.deviceId = uuid
        }

        engine!!.update(device2)

        engine!!.getAll(TABLE_NAME_SINGLE_ENTRY, { savedDevice = it }).join()
        assertEquals(1, savedDevice!!.size.toLong())
        assertEquals(device.deviceId, savedDevice!![0].deviceId)
    }

    @After
    fun tearDown() {
        engine!!.close()
        encryptedEngine!!.close()
    }

    companion object {

        private val ENCRYPTION_KEY = "custom_key"
    }
}
